### ISSUE TITLE
cxtype_ref problem for some gcc versions

### ISSUE THREAD

**Issue by roiser** [madgraph5/madgraph4gpu#1004]
I stumbled over a slight problem when running with some (sic!) gcc versions (e.g. 12.4, 13.3 or 14.2, didn't try <= 11 series). I.e. a compilation error of the code 

```	    ../../src/mgOnGpuVectors.h: In member function ‘mg5amcCpu::cxtype_ref mg5amcCpu::cxtype_v::operator[](size_t) const’:
	    ../../src/mgOnGpuVectors.h:108:74: error: binding reference of type ‘mgOnGpu::fptype&’ {aka ‘double&’} to ‘const double’ discards qualifiers
	      108 |     cxtype_ref operator[]( size_t i ) const { return cxtype_ref( m_real[i], m_imag[i] ); }
	          |                                                                  ~~~~~~~~^
	    In file included from ../../src/mgOnGpuVectors.h:9:
	    ../../src/mgOnGpuCxtypes.h:716:45: note:   initializing argument 1 of ‘mg5amcCpu::cxtype_ref::cxtype_ref(mgOnGpu::fptype&, mgOnGpu::fptype&)’
	      716 |     __host__ __device__ cxtype_ref( fptype& r, fptype& i )
	          |                                     ~~~~~~~~^ 
```

which seems correct as the ```operator[]``` is ```const``` and then discards the constness when calling ```cxtype_ref``` which has no overload with const arguments. 

As this special operator was guarded by a macro ```MGONGPU_HAS_CPPCXTYPEV_BRK``` I looked into this direction and found the special gcc implementation (not used e.g. for clang). Undefining the macro then makes the code work again. 

The funny (haha) thing about it is that as mentioned above this behavior seems to be only true for some distinct gcc versions (12.4, 13.3, 14.2) versions inbtetween those work !!!! I made a reproducer of the problem at https://godbolt.org/z/oM8v8Wo66 . One thing I haven't checked yet is whether disabling the operator with the "inbetween" versions nowadays works.

**Comment by valassi** [comment]
Thanks @roiser I can look at it this week

**Comment by roiser** [comment]
i just recompiled with gcc 11.4 and this also works __without__ the specialised bracket operator

**Comment by valassi** [comment]
Thanks @roiser I confirm that 14.1 is ok but 14.2 fails (I have no access to 13.3 or 12.4 but will focus on 14.2)
```
[avalassi@itscrd90 gcc14.2/cvmfs] /data/avalassi/GPU2023/madgraph4gpuX/epochX/cudacpp/gg_tt.mad/SubProcesses/P1_gg_ttx> make -f cudacpp.mk 
BACKEND=cpp512y (was cppauto)
OMPFLAGS=
FPTYPE='d'
HELINL='0'
HRDCOD='0'
HASCURAND=hasNoCurand
HASHIPRAND=hasNoHiprand
Building in BUILDDIR=. for tag=512y_d_inl0_hrd0_hasNoCurand_hasNoHiprand (USEBUILDDIR != 1)
make -C ../../src  -f cudacpp_src.mk
make[1]: Entering directory '/data/avalassi/GPU2023/madgraph4gpuX/epochX/cudacpp/gg_tt.mad/src'
ccache /cvmfs/sft.cern.ch/lcg/releases/gcc/14.2.0-2f0a0/x86_64-el9/bin/g++  -I. -O3  -std=c++17 -Wall -Wshadow -Wextra -ffast-math   -march=skylake-avx512 -mprefer-vector-width=256  -DMGONGPU_FPTYPE_DOUBLE -DMGONGPU_FPTYPE2_DOUBLE -fPIC -DMGONGPU_CHANNELID_DEBUG -c read_slha.cc -o read_slha_cpp.o
ccache /cvmfs/sft.cern.ch/lcg/releases/gcc/14.2.0-2f0a0/x86_64-el9/bin/g++  -I. -O3  -std=c++17 -Wall -Wshadow -Wextra -ffast-math   -march=skylake-avx512 -mprefer-vector-width=256  -DMGONGPU_FPTYPE_DOUBLE -DMGONGPU_FPTYPE2_DOUBLE -fPIC -DMGONGPU_CHANNELID_DEBUG -c Parameters_sm.cc -o Parameters_sm_cpp.o
In file included from Parameters_sm.h:21,
                 from Parameters_sm.cc:15:
mgOnGpuVectors.h: In member function 'mg5amcCpu::cxtype_ref mg5amcCpu::cxtype_v::operator[](size_t) const':
mgOnGpuVectors.h:108:74: error: binding reference of type 'mgOnGpu::fptype&' {aka 'double&'} to 'const double' discards qualifiers
  108 |     cxtype_ref operator[]( size_t i ) const { return cxtype_ref( m_real[i], m_imag[i] ); }
      |                                                                  ~~~~~~~~^
In file included from Parameters_sm.h:20:
mgOnGpuCxtypes.h:716:45: note:   initializing argument 1 of 'mg5amcCpu::cxtype_ref::cxtype_ref(mgOnGpu::fptype&, mgOnGpu::fptype&)'
  716 |     __host__ __device__ cxtype_ref( fptype& r, fptype& i )
      |                                     ~~~~~~~~^
make[1]: *** [cudacpp_src.mk:135: Parameters_sm_cpp.o] Error 1
make[1]: Leaving directory '/data/avalassi/GPU2023/madgraph4gpuX/epochX/cudacpp/gg_tt.mad/src'
make: *** [cudacpp.mk:771: ../../lib/libmg5amc_common_cpp.so] Error 2
```

**Comment by roiser** [comment]
> Thanks @roiser I can look at it this week

No worries, I can also followup

**Comment by valassi** [comment]
This is enough: distinguish between const and non-const operator[] in cxtype_v
```
diff --git a/epochX/cudacpp/gg_tt.mad/src/mgOnGpuVectors.h b/epochX/cudacpp/gg_tt.mad/src/mgOnGpuVectors.h
index 1693fcea7..e28a7ff5f 100644
--- a/epochX/cudacpp/gg_tt.mad/src/mgOnGpuVectors.h
+++ b/epochX/cudacpp/gg_tt.mad/src/mgOnGpuVectors.h
@@ -105,7 +105,9 @@ namespace mg5amcCpu
     // NB: THIS IS THE FUNDAMENTAL DIFFERENCE BETWEEN MGONGPU_HAS_CPPCXTYPEV_BRK DEFINED AND NOT DEFINED
     // NB: the alternative "clang" implementation is simpler: it simply does not have any bracket operator[]
     // NB: ** do NOT implement operator[] to return a value: it does not fail the build (why?) and gives unexpected results! **
-    cxtype_ref operator[]( size_t i ) const { return cxtype_ref( m_real[i], m_imag[i] ); }
+    //cxtype_ref operator[]( size_t i ) const { return cxtype_ref( m_real[i], m_imag[i] ); } // gcc14.2 build fails #1004
+    cxtype_ref operator[]( size_t i ) { return cxtype_ref( m_real[i], m_imag[i] ); }
+    cxtype operator[]( size_t i ) const { return cxtype( m_real[i], m_imag[i] ); }
 #endif
     const fptype_v& real() const
     {
```

**Comment by roiser** [comment]
Hi @valassi I understand that you want to investigate this issue, please go ahead, thanks Stefan

**Comment by valassi** [comment]
Thanks @roiser a fix in in PR #1006

**Comment by valassi** [comment]
> i just recompiled with gcc 11.4 and this also works **without** the specialised bracket operator

Sorry I forgot to answer on this, thanks for the observation

Yes the 'alternative' implementation that I had developed for clang would also work for gcc
```
diff --git a/epochX/cudacpp/gg_tt.mad/src/mgOnGpuVectors.h b/epochX/cudacpp/gg_tt.mad/src/mgOnGpuVectors.h
index 9f3533a87..25664eadb 100644
--- a/epochX/cudacpp/gg_tt.mad/src/mgOnGpuVectors.h
+++ b/epochX/cudacpp/gg_tt.mad/src/mgOnGpuVectors.h
@@ -27,8 +27,8 @@
 //#define MGONGPU_HAS_CPPCXTYPEV_BRK 1 // icc default?
 #undef MGONGPU_HAS_CPPCXTYPEV_BRK // icc test
 #else
-#define MGONGPU_HAS_CPPCXTYPEV_BRK 1 // gcc default
-//#undef MGONGPU_HAS_CPPCXTYPEV_BRK // gcc test (very slightly slower? issue #172)
+//#define MGONGPU_HAS_CPPCXTYPEV_BRK 1 // gcc default
+#undef MGONGPU_HAS_CPPCXTYPEV_BRK // gcc test (very slightly slower? issue #172)
 #endif
```

This is something I had looked in April 2021 in #172 and specifically here
https://github.com/madgraph5/madgraph4gpu/issues/172#issuecomment-827711862

> About the second issue, whether the clang version can be used in production also for gcc, this is now validated. On ecould use that version. However, it is very tinily slower (a few permille). And I lile the original operator[] idea. I will keep as is for the moment.

By now I think that I do not care about he permille slowdown. But I still like the idea of keeping two parallel implementations, just in case there is a problem in one of them (as there is now). Also because it gives interesting understandings of compiler vector extensions, which are the central tool in pur SIMD code. So for the moment I would keep the brackets for gcc and the no-brackets in clang. Then we can always reassess

Thanks again @roiser for the report and the suggetsions

**Comment by valassi** [comment]
This is fixed in PR #1006. Closing
